avalon/photoshop/lib.py

```diff
@@ -0,0 +1,255 @@
+# or Mac.
+# This needs to be a partial function we can later call because when calling
+# Dispatch directly from a different thread will result first a
+# "CoInitialize has not been called.", which can be fixed with
+# pythoncom.CoInitialize() but secondly "The application called an interface
+# that was marshalled for a different thread".
+app = functools.partial(Dispatch, "Photoshop.Application")
+
+
+def start_server():
+    """Starts the web server that will be hosted in the Photoshop extension"""
+    from avalon import api, photoshop
+
+    api.install(photoshop)
+    html_server.app.start_server(5000)
+
+
+def imprint(layer, data):
+    """Write `data` to the active document "headline" field as json.
+
+    Arguments:
+        layer (win32com.client.CDispatch): COMObject of the layer.
+        data (dict): Dictionary of key/value pairs.
+
+    Example:
+        >>> from avalon.photoshop import lib
+        >>> layer = app.ActiveDocument.ArtLayers.Add()
+        >>> data = {"str": "someting", "int": 1, "float": 0.32, "bool": True}
+        >>> lib.imprint(layer, data)
+    """
+    layers_data = {}
+    try:
+        layers_data = json.loads(app().ActiveDocument.Info.Headline)
+    except json.decoder.JSONDecodeError:
+        pass
+
+    # json.dumps writes integer values in a dictionary to string, so
+    # anticipating it here.
+    if str(layer.id) in layers_data:
+        layers_data[str(layer.id)].update(data)
+    else:
+        layers_data[str(layer.id)] = data
+
+    app().ActiveDocument.Info.Headline = json.dumps(layers_data, indent=4)
+
+
+def read(layer):
+    """Read the layer metadata in to a dict
+
+    Args:
+        layer (win32com.client.CDispatch): COMObject of the layer.
+
+    Returns:
+        dict
+    """
+    layers_data = {}
+    try:
+        layers_data = json.loads(app().ActiveDocument.Info.Headline)
+    except json.decoder.JSONDecodeError:
+        pass
+
+    return layers_data.get(str(layer.id))
+
+
+@contextlib.contextmanager
+def maintained_selection():
+    """Maintain selection during context."""
+    selection = get_selected_layers()
+    try:
+        yield selection
+    finally:
+        select_layers(selection)
+
+
+@contextlib.contextmanager
+def maintained_visibility():
+    """Maintain visibility during context."""
+    visibility = {}
+    layers = get_layers_in_document(app().ActiveDocument)
+    for layer in layers:
+        visibility[layer.id] = layer.Visible
+    try:
+        yield
+    finally:
+        for layer in layers:
+            layer.Visible = visibility[layer.id]
+
+
+def group_selected_layers():
+    """Create a group and adds the selected layers."""
+
+    ref = Dispatch("Photoshop.ActionReference")
+    ref.PutClass(app().StringIDToTypeID("layerSection"))
+
+    lref = Dispatch("Photoshop.ActionReference")
+    lref.PutEnumerated(
+        app().CharIDToTypeID("Lyr "),
+        app().CharIDToTypeID("Ordn"),
+        app().CharIDToTypeID("Trgt")
+    )
```

Comment:
Is it required to initiliaze the Dispatch() each time using `app()`? Or could you do it once and trigger all the commands afterwards?

```python
app = get_app()
app.CharIDToTypeId("Lyr ")
app.CharIDToTypeId("Ordn")
app.CharIDToTypeId("Trgt")
```

No idea how the interface works, but just wondering.

Reply:
Yeah, we could certainly do that within functions.